### PR TITLE
Assignment2 vibrato peer review

### PR DESCRIPTION
**Pull request by SileYin**

### CODE REVIEW

File: src/Vibrato/Vibrato.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    m_fSampleRate = fSampleRateInHz;
+    m_fDelayTimeInS = fDelayTimeInS;
+
+    int iMaxDelayLength = static_cast<int>(2 * std::round(m_fDelayTimeInS * m_fSampleRate)) + 2;
+
+    m_pCLFO = new CWavetableLFO(iLFOWavetableLength, m_fSampleRate, m_fVibratoFrequencyInHz, m_fVibratoRangeInS * m_fSampleRate);
+    m_pCRingBuff = new CRingBuffer<float>(iMaxDelayLength);
+    m_pCRingBuff->setReadIdx(static_cast<int>(std::round(m_fDelayTimeInS * m_fSampleRate)) + 2);
```
Comment (nol-alb) [inline src/Vibrato/Vibrato.cpp:50]: So you are setting the readIdx at the end of the delay line to avoid an initial delay in the output? Or did i get this wrong, this is really interesting!

File: src/Tests/Tests/Test_Vibrato.cpp [hunk 2]
```diff
@@ -3,19 +3,331 @@
...
 #include "gtest/gtest.h"
-
+#include "Synthesis.h"
 
 namespace vibrato_test {
     void CHECK_ARRAY_CLOSE(float* buffer1, float* buffer2, int iLength, float fTolerance)
     {
         for (int i = 0; i < iLength; i++)
         {
+            float error = abs(buffer1[i] - buffer2[i]);
```
[the remark this answers is not in the document]
Comment (SileYin) [reply src/Tests/Tests/Test_Vibrato.cpp:15]: This is a debug code that I forgot to get rid of, should be deleted

File: src/Vibrato/Vibrato.cpp (hunk 1 above)
Comment (SileYin) [reply src/Vibrato/Vibrato.cpp:50]: This is a trick to set a fixed delay line. Same result could be achieved when you initialize with the write head delayed

**Review by nol-alb — COMMENTED** [review]
Really good implementation, interesting ideas, I learnt a bit from this code and style thank you!

File: src/inc/Lfo.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /*! Set the phase for the read head
+    \param fPhaseInDegrees phase for the read head in degrees
+    */
+    void setReadPhase(float fPhaseInDegrees)
+    {
+        while (fPhaseInDegrees >= 360)
+            fPhaseInDegrees -= 360;
+        while (fPhaseInDegrees < 360)
```
[the remark this answers is not in the document]
Comment (SileYin) [reply src/inc/Lfo.h:44]: Yes it should be. My mistake here, will fix it later. However it won't cause an error since this would be automatically wrapped back in the ringbuffer get function.

File: src/MUSI6106Exec/MUSI6106Exec.cpp [hunk 4]
```diff
@@ -15,12 +17,118 @@ void    showClInfo ();
...
+    ppfAudioDataOut = new float*[stFileSpec.iNumChannels];
+    for (int i = 0; i < stFileSpec.iNumChannels; i++)
+        ppfAudioDataOut[i] = new float[kBlockSize];
+    
+    
+    CVibrato::create(pCVibrato);
+    pCVibrato->init(fVibratoRange, stFileSpec.fSampleRateInHz);
+    pCVibrato->setParam(CVibrato::kParamVibratoFrequency, fVibratoFrequency);
+    pCVibrato->setParam(CVibrato::kParamVibratoRange, fVibratoRange);
+
```
[the remark this answers is not in the document]
Comment (RoseQSun) [reply src/MUSI6106Exec/MUSI6106Exec.cpp:92]: I think Noel raised a valid point. However, I understand this design choice as it is consistent in style with CCombFilterIf.

File: src/Tests/Tests/Test_Vibrato.cpp [hunk 5]
```diff
@@ -3,19 +3,331 @@
...
+    {
+        for (int i = 0; i < 5; i++)
+        {
+            pCRingBuff->putPostInc(1.F * i);
+        }
+
+        for (int i = 5; i < 30; i++)
+        {
+            EXPECT_EQ(pCRingBuff->getNumValuesInBuffer(), 5);
+            EXPECT_NEAR(pCRingBuff->getPostInc(), i - 5, 1e-3);
```
Comment (RoseQSun) [inline src/Tests/Tests/Test_Vibrato.cpp:71]: Would you explain the reason why you used NEAR instead of EQ here and in the following Ring Buffer tests?

**Comment by RoseQSun** [comment]
Program works without errors and passes required tests (Those comparison results were impressive). Code is clean with proper comments (I particularly enjoyed the doxygen implementation). Good job guys! 
Some nitpicking: As suggested in the assignment requirements, you could test your system by setParam with various values (e.g., negative parameters). You could make use of Synthesis.h in your testing and lfo generation if you'd like, but the current method works really well.

File: src/Tests/Tests/Test_Vibrato.cpp (hunk 5 above)
Comment (SileYin) [reply src/Tests/Tests/Test_Vibrato.cpp:71]: For floating point numbers using equal makes no sense

**Comment by Jiarui-Xu-Gatech** [comment]
**In Vibrato.cpp**

1.Error_t CVibrato::process(float* pfInputBuffer, float* pfOutputBuffer, int iNumberOfFrames)

**I would recommend to use float** ppfInputBuffer instead of float* pfInputBuffer, just in case the input can be two channels, so that one dimension of ppfInputBuffer can be used to represent the index of channel.**

**In Lfo.h**

2.(line 20) CWavetableLFO(int iWavetableLength, int iSampleRate, float fFrequency, float fAmplitude):
        m_iSampleRate(iSampleRate),
        m_dReadIdx(0)
    {
        double const l_dPi = 3.14159265358979323846;
        assert(iWavetableLength > 0);
        pCWavetableRingBuff = new CRingBuffer<float>(iWavetableLength);
        for (int i = 0; i < iWavetableLength; i++)
        {
            pCWavetableRingBuff->putPostInc(sin(2 * l_dPi * i / iWavetableLength));
        }

**I wonder why you use double const l_dPi = 3.14159265358979323846. In fact, math.h's M_PI is available to use.**

3.(line 31) ~CWavetableLFO()
    {
        delete pCWavetableRingBuff;
        pCWavetableRingBuff = 0;
    }

**If you want to delete pCWavetableRingBuff, why don't you use delete[]. In this way, are you sure that pCWavetableRingBuff only locate in one stuck of memory? Or, if you just want to change pCWavetableRingBuff's value to 0, why do you delete it?**

**In MUSI6106Exec.cpp**

4.(line 29)float **ppfAudioData = 0;
    float **ppfAudioDataOut = 0;

**So now I know that you really put double-pointer in case the input is two-channel. Great**

5.(line 94)while (!phAudioFile->isEof())
    {
        // set block length variable
        long long iNumFrames = kBlockSize;

        // read data (iNumOfFrames might be updated!)
        phAudioFile->readData(ppfAudioData, iNumFrames);
        //cout << "\r" << "reading audio"<<endl;
        
        // vibrato
        pCVibrato->process(ppfAudioData[0], ppfAudioDataOut[0], iNumFrames);
        
        //cout << "\r" << "Write audio"<<endl;
        phAudioFileOut->writeData(ppfAudioDataOut, iNumFrames);
        
    }

**I just wonder how you break this while...... Does pCVibrato->process influence phAudioFile->isEof()'s result?**

**In Test_Vibrato.cpp**

6.(line 171)TEST_F(Vibrato, DCInputEqualOutput)
    {
        pCVibrato->init(0.01, 48000);
        pCVibrato->setParam(CVibrato::kParamVibratoFrequency, 10);
        pCVibrato->setParam(CVibrato::kParamVibratoRange, 0.005);
        pfInputBuffer = new float[2048];
        pfOutputBuffer = new float[2048];
        for (int i = 0; i < 2048; i++)
        {
            pfInputBuffer[i] = 1;
        }
        pCVibrato->process(pfInputBuffer, pfOutputBuffer, 2048);
        CHECK_ARRAY_CLOSE(pfInputBuffer + 240, pfOutputBuffer + 240 + 480, 2048 - 480 - 240, 1e-3);

        pCVibrato->setParam(CVibrato::kParamVibratoFrequency, 20);
        pCVibrato->setParam(CVibrato::kParamVibratoRange, 0.002);
        pCVibrato->process(pfInputBuffer, pfOutputBuffer, 2048);
        CHECK_ARRAY_CLOSE(pfInputBuffer + 96, pfOutputBuffer + 96 + 480, 2048 - 480 - 96, 1e-3);


        delete[] pfInputBuffer;
        delete[] pfOutputBuffer;
    }

**I don't think it is a good idea to use so much hard code there. It is difficult to modify and also not friendly to readers. I recommend to use variable instead of hard code to represent the number of parameters.


The code overall is good without problems, just some small things to modify. Thank you.**

**Review by medioqrity — APPROVED** [review]
I have carefully read all changed files and does not find any possible bugs or design deficiencies. The tests are well written and covers edge cases. Codes are written in good structure and well commented. Good job.

File: src/Vibrato/Vibrato.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+float CVibrato::getParam(VibratoParam_t eParam) const
+{
+    if (!m_bIsInitialized)
+    {
+        return -1;
+    }
+    if (eParam == CVibrato::kParamVibratoFrequency)
+    {
+        return m_pCLFO->getLFOFrequency();
```
Comment (ijc8) [inline src/Vibrato/Vibrato.cpp:101]: Given that CVibrato can just set/get the parameters in the LFO, as you're doing here, does it need to have the information duplicated as member variables (`m_fVibratoRangeInS`, `m_fVibratoFrequencyInHz`)?

File: src/inc/Lfo.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    \param iWavetableLength wavetable length of the LFO
+    \param iSampleRate sample rate for the LFO
+    \param fFrequency frequency for the LFO in Hz
+    \param fAmplitude amplitude for the LFO
+    */
+    CWavetableLFO(int iWavetableLength, int iSampleRate, float fFrequency, float fAmplitude):
+        m_iSampleRate(iSampleRate),
+        m_dReadIdx(0)
+    {
+        double const l_dPi = 3.14159265358979323846;
```
Comment (ijc8) [inline src/inc/Lfo.h:20]: I was going to ask why not just use `M_PI`, but apparently that's actually a nonstandard extension (although it is part of [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/math.h.html)). Weird!
Anyway, apparently C++20 [finally includes](https://en.cppreference.com/w/cpp/numeric/constants) a constant for `pi`, so maybe someday we can collectively stop rewriting this line of code. :stuck_out_tongue:

File: src/inc/Lfo.h [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    \param fPhaseInDegrees phase for the read head in degrees
+    */
+    void setLFOFrequency(float fFrequency, float fPhaseInDegrees = (0.f))
+    {
+        while (fPhaseInDegrees >= 360)
+            fPhaseInDegrees -= 360;
+        while (fPhaseInDegrees < 360)
+            fPhaseInDegrees += 360;
+        m_dReadInc = static_cast<double>(pCWavetableRingBuff->getLength()) / static_cast<double>(m_iSampleRate) * fFrequency;
+        setReadPhase(fPhaseInDegrees);
```
Comment (ijc8) [inline src/inc/Lfo.h:68]: What's the motivation for resetting the phase when updating the frequency? It seems like this would prevent smoothly varying the vibrato frequency.

File: src/inc/Lfo.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    \param fFrequency frequency for the LFO in Hz
+    \param fAmplitude amplitude for the LFO
+    */
+    CWavetableLFO(int iWavetableLength, int iSampleRate, float fFrequency, float fAmplitude):
+        m_iSampleRate(iSampleRate),
+        m_dReadIdx(0)
+    {
+        double const l_dPi = 3.14159265358979323846;
+        assert(iWavetableLength > 0);
+        pCWavetableRingBuff = new CRingBuffer<float>(iWavetableLength);
```
Comment (ijc8) [inline src/inc/Lfo.h:22]: Does this need to be a pointer allocated on the heap? If the member variable was just a `CRingBuffer`, rather than a pointer, you could set it up in the initializer list and not have to worry about remembering to `delete` it.
(This is possible `CWavetableLFO` because it does its initialization in the constructor, unlike `CVibrato`.)

File: src/inc/Vibrato.h [hunk 10]
```diff
@@ -1,4 +1,95 @@
...
+* don't have much value. If the user wants multichannel vibrato, they can just 
+* have one instance of this class for each channels.
+*  
+* This interface is designed (100% copied) from the previous CCombFilterIf and
+* AudioFileIf classes to keep the repository in style.
+* The create and destroy function and the protected constructor and destructor
+* are probably designed to provide extra memory management to the user. However, 
+* I haven't looked into when the program throw an error and run out of the scope
+* of an instance, would this give memory issue or not. Modern C++ has things like
+* smart pointers to manage memories, don't know if it would be better to use that.
```
Comment (ijc8) [inline src/inc/Vibrato.h:21]: Destructors should still get called regardless of how you exit the scope.
Smart pointers are nice; `std::unique_ptr`, for instance, makes it harder to accidentally leak memory by tying allocation/deallocation to scope (as occurs automatically for variables on the stack).
IMO, it's even better if you can avoid unnecessary pointers (and thus indirection + potential leaks) altogether. In this case, there is no real need for the LFO and RingBuffer to be allocated separately on the heap; the only reason is because the Vibrato's construction and initialization have been split up (into the constructor and a separate `init()` method). (This also means CVibrato can be in an awkward "created but unusable" state which doesn't afflict your CWavetableLFO.)
But I appreciate that you're following the conventions of other classes in the repo.

**Review by ijc8 — COMMENTED** [review]
Thanks for keeping this PR clean and basing it on the starter code! It makes reviewing easier. :slightly_smiling_face:

File: src/Vibrato/Vibrato.cpp (hunk 6 above)
Comment (SileYin) [reply src/Vibrato/Vibrato.cpp:101]: Decisions around this is a little bit hard to make since sometimes an explicit member function would just make life easier but sometimes I also want to avoid duplicates. I will revisit this and probably change it to make it cleaner and more consistent around the codebase!

File: src/inc/Lfo.h (hunk 8 above)
Comment (SileYin) [reply src/inc/Lfo.h:68]: Good point, looks like this is a little bit unnecessary and would have potential artifacts, will change it later.

**Comment by nitin-hugar** [comment]
I think the code has been well written and clear and I was able to build and run the program and tests without issues. The comments made it easy to understand your choices. Although, as Ian and Jiarui have mentioned, some hardcoding of values could have been avoided. Overall, good job.
